app/components/Wallet/SuiWallet/useBalance.tsx

```diff
@@ -23,21 +26,48 @@ export function useCoinBalance(coinOrVariant?: string) {
 				? nbtc.pkgId + nbtc.coinType
 				: coinOrVariant;
 
+	const coinCacheKey = ["coinBalance", userAddr, network, coinOrVariant || "SUI"];
+
 	const {
 		data: balance,
 		error,
 		isLoading,
 		refetch,
-	} = useQuery({
-		queryKey: ["coinBalance", userAddr, network, resolvedCoinAddr],
+	} = useQuery<CoinBalance>({
+		queryKey: coinCacheKey,
 		queryFn: () => suiClient.getBalance({ owner: userAddr!, coinType: resolvedCoinAddr }),
 		enabled: !!userAddr,
+		// cache the balance for 10 min
+		staleTime: 600000,
 	});
 
+	const updateCoinBalanceInCache = (newBalance: string) => {
+		queryClient.setQueryData(coinCacheKey, (oldData: CoinBalance) => ({
+			...oldData,
+			totalBalance: newBalance,
+		}));
+	};
+
 	return {
 		balance: balance ? BigInt(balance.totalBalance) : 0n,
 		isLoading,
 		error,
 		refetch,
+		updateCoinBalanceInCache,
 	};
 }
+
+// observe balance changes and update the cache for the corresponding coin tpye
+export function handleBalanceChanges(
+	balanceChanges: BalanceChange[],
+	coinType: string,
+	currentBalance: bigint,
+	updateCoinBalanceInCache: (newBalance: string) => void,
+) {
+	for (const balanceChange of balanceChanges) {
+		if (balanceChange.coinType === coinType) {
+			const newBalance = String(currentBalance + BigInt(balanceChange.amount));
+			updateCoinBalanceInCache(newBalance);
```

Comment:
you should call it once at the end of the loop, and only if the newBalance is different than the old one.